### DISCUSSION TITLE
Inline worker in webpack 5

### DISCUSSION THREAD

**Discussion by lstkz** [webpack/webpack#14066] [Q&A]
I need to host all assets in CDN, but Web Worker requires the same domain policy.

The below code works correctly if I don't use CDN.

```ts
    this.worker = new Worker(
      new URL('./BundlerService.worker.ts', import.meta.url)
    );
```

Here are the docs https://webpack.js.org/guides/web-workers/, but they don't mention how to make inline workers.

worker-loader for webpack4 https://github.com/webpack-contrib/worker-loader#cross-origin-policy

> WebWorkers are restricted by a same-origin policy, so if your webpack assets are not being served from the same origin as your application, their download may be blocked by your browser. This scenario can commonly occur if you are hosting your assets under a CDN domain. Even downloads from the webpack-dev-server could be blocked.
> There are two workarounds:
> Firstly, you can inline the worker as a blob instead of downloading it as an external script via the inline parameter

In the old approach, there is a workaround for it, but I can't find any solutions for wepback5.

I know, it's possible to use a proxy from the same domain, but I need to use CDN because of the performance.

**Comment by alexander-akait** [comment]
In theory we can implement `assert/blob`, but here other edge cases, browsers without blob supporting, but it is fixable using `asset/inline`

**Reply by lstkz** [reply]
https://caniuse.com/?search=blob it seems to be supported in all browser

**Reply by alexander-akait** [reply]
Don't forget about outdated browser, but yes, we can implement this and some limitations about old browsers versions

**Reply by lstkz** [reply]
Fine for me, thanks!

**Reply by thexeos** [reply]
`asset/inline` should be sufficient for our use case. If separate chunk is required, the file where Worker is defined can be imported asynchronously at runtime, creating a new chunk.

**Reply by MikeChongCan** [reply]
Sorry, could you elaborate how to use `asset/inline` with Webpack 5 for a worker?

**Reply by AprilArcus** [reply]
Echoing @imWildCat. We are attempting to migrate from `worker-loader`, but our assets are served by CDN and are running afoul of `new Worker()`'s same-origin policy.

**Reply by AprilArcus** [reply]
Something like this?

```
module: {
  rules: [
    // ...
    {
      resourceQuery: /inline/,
      type: 'asset/inline',
    }
    // ...
  ]
},
```

and then this?

`new Worker(new URL('./relative/path/to/my/worker.js?inline'), import.meta.url)`

Will this break in node/jest when the query string is reached?

**Reply by AprilArcus** [reply]
is there a way to directly test in `module.rules` for a request coming from the `new Worker()` constructor?

**Reply by mmazzarolo** [reply]
@AprilArcus , related: https://github.com/webpack/webpack/discussions/14198

**Reply by johannesodland** [reply]
Did anyone manage to find a workaround to build and load inline workers? We stumbled upon the same issue.

**Reply by thexeos** [reply]
> Did anyone manage to find a workaround to build and load inline workers? We stumbled upon the same issue.

0) If you need WASM in the worker, first add this to webpack config:

```js
cfg.module.rules.unshift({
  test: /\.wasm/,
  type: 'asset/inline'
})
```

1.a) If this is a [static] third-party worker (like QR code reader or audio processor), then add it to public directory and reference it via full path. Code would look like `new Worker('/js/static-worker-file-name.js')`.

1.b) If you are actively working on the worker code, you need to use [`workerize-loader`](https://github.com/developit/workerize-loader) until webpack releases this functionality. Install `workerize-loader` via package manager and add the following code to the file where you would otherwise have `new Worker()` code:

```
// eslint-disable-next-line
import createWorker from 'workerize-loader?inline!./worker-file.worker.js'
```

_Note: you may or may not need the eslint line (depending on your webpack config) and you may or may not need `.js` extension. If the worker file is in a different directory all together, then the import string would look like `workerize-loader?inline!../../some-other-folder/another-folder/worker-file.worker.js`. Again, adapt to your project._

Then, where you would need a `Worker` instance, run `createWorker()` (run it once and save the returned value in a variable).

_Note: as a side-effect, there may be a randomly named file appearing in the root of your "dist" folder with contents of the `workerize-loader` output and with `.worker.js` extension. In my case, this is a "bug" - the same file is already being inlined in the JS bundle, however you may or may not need it in your project depending on your webpack config. Thus, you should test it: check Network tab of Dev Tools to see if that file is being loaded, and if not, then you can discard it from your build using a webpack plugin (or just ignore it)._

**Comment by Danielku15** [comment]
I am also currently facing challenges in regards to WebPack and WebWorkers (+AudioWorklets). While I [could figure out](https://github.com/webpack/webpack/discussions/15382) most of my aspects, exactly the point of having inline workers is still open. So I'm joining here the discussion related to the support of inline workers. 

The  `new Worker(new Url('worker-module', import.meta.url))` detection seems to work for some cases, but it is not optimal for two cases: 

1. You deliver your workers from a different origin (CDN) - Discussed here already
2. It only handles Web Workers but not Audio Worklets. - Not sure if we can solve this case together with the first one. Might be off-topic for this discussion. 

**Case 1:** I would vote for a new syntax which triggers the same worker generation logic like today, but not being dependent on the `new Worker` syntax. This would allow developers to obtain the URL to an auto generated worker and use it as they see fit. 
_Pro:_ Developers can decide themselves how to spawn workers using the available browser APIs and other configurations they might have. This keeps the complexity out of WebPack to stay up to date about available practices. 
_Con:_ I wouldn't be aware of a good syntax which is part of the standard browser APIs. A WebPack specific syntax implies some risks but could be a temporary solution at least. Maybe a new [module variable](https://webpack.js.org/api/module-variables/) like `__webpack_get_worker_url__('worker-module')` could be added?

**Comment by Danielku15** [comment]
@alexander-akait As per my proposal I started with a change which would introduce a new module variable. I have it over at my fork https://github.com/Danielku15/webpack/tree/feature/inline-workers ([diff here](https://github.com/webpack/webpack/compare/main...Danielku15:feature/inline-workers)). It basically consists of following changes: 

1. A new module variable (actually function 😁 ) `__webpack_get_worker_url__(..)` which is treated similar like a `new Worker(..)`. 
2. An implementation for `__webpack_get_worker_url__` which simply returns the URL of the worker.

The trickiest part was to get it running in the test environment because the whole worker part is customized in the WebPack worker environment to run in a Node environment. I had to fiddle a bit to get the tests working and I couldn't make a full test because I ran into strange syntax errors being reported if I used imports in the inline workers.  

I didn't dare yet to open a PR because we don't have confirmation yet if the additional module variable would be your preferred solution to this problem. But it gives an idea. The tests give a good indication about how it would be used: 

https://github.com/webpack/webpack/compare/main...Danielku15:feature/inline-workers#diff-078d5ff978b3c8f1c6b934494ed6fd46b6b7bc98c4375ddb96ef579dea79de53R1

If you are fine with this path, I could open a PR for a more detailed discussion on the code changes and what further tests should be added. 

If this change finds its way into WebPack, I would try to also build on top of this a worklet support (https://github.com/webpack/webpack/issues/11543). For this I would then need to adapt the built-in `importScripts` which is used within the generated worker file.

**Reply by alexander-akait** [reply]
@vankop What do you think?

**Reply by AprilArcus** [reply]
What about having a compiler-wide config to build all workers inline or not? In my case there is no chance that some workers will be CDN and others local; it's all or nothing.

I'd prefer this to a special syntax as "new Worker" is valuable for NodeJS isomorphic builds.

**Reply by Danielku15** [reply]
For me it would be rather a non-goal to get the workers inlined by WebPack itself 🤔 . It does not really solve all use-cases I am trying to address by the proposed change. 

1. I have a library which spawns itself also as worker. Within my library I would like  to use `__webpack_get_worker_url__` to opt in for an inline worker approach for my library. My library is not specific to WebPack but can detect whether it was bundled using it, and then reacts accordingly. Compiler wide configs would imply that I am the integrator of the library (developer of the final application) but that's not the case for me, I am providing a JS file which can be bundled with WebPack, but doesn't have to. I would need to tell all people in the market to change their config to inline workers just because my library would require it. This is not in my interest.
2. We would still need to develop a second solution for the worklet support with additional detection mechanisms when to generate worker code for them and then provide the correct URL there. https://github.com/webpack/webpack/issues/11543 Having manual access to the URL makes it maybe a bit more verbose to use WebPack in this context, but it also makes the different usecases of people easier to solve, because they can decide themselves how they use it further 
  * Use the URL to create blobs with inline source code using imports. (and injecting additional overrides through already available webpack customizations like `__webpack_chunk_load__ `)
  * Use the URL to create blobs with inline source code using importScripts (and injecting additional overrides through already available webpack customizations like `__webpack_chunk_load__ `)
  * Pass on the URL to any other function which might create a worker or worklet (`addModule`)

The `new Worker()` approach requires developers to always exactly write the correct syntax which WebPack supports to achieve the right behavior. And also WebPack might only support some very specific scenarios of directly spawning the workers or configuring the worklet at that place. The more dynamic your code is, the more likely WebPack will fail to detect the _right_ approach (create wrapper, provide URL to it) it should take and will fallback to a different approach silently.

**Reply by vankop** [reply]
I lost context a little bit..
So you trying to achieve:
- load web worker code from cdn
- create blob
- create web worker from blob

am I right?

**Reply by Danielku15** [reply]
I hope this message will not get again too long, but to bring you on-board regarding my needs I need to explain a bit more 😅 

On high level you are right, but there are some details to that which are important to understand the full use-case: 

**The CDN Use Case is:** 
* ✅  We have an application/library which spawns a WebWorker using the `new Worker()` syntax. 
* ✅ It is bundled with WebPack which initiates the Worker Wrapper generation and it places the respective relative URL into the output code. 
* ✅ The application/library and the worker file are all uploaded to CDN (cdn01.mycompany.com). 
* ✅ We have a website which includes the files from the CDN but is hosted on a different domain. This application might also have code bundled with WebPack (in the library use-case)
* ✅ The application/script executes and tries to spawn the worker.
* .⚠️ The browser will block the loading of the worker because the website origin (app.mycompany.com) is not allowed to spawn a worker with the absolute url like cdn01.company.com/worker.js

**Web Workers (and Worklets) have the following rules:** 
* You cannot spawn a worker with a URL which is on a different origin. 
* A worker spawned with the current origin, can load any scripts and resources from other origins following the CORS policies defined through headers. 

_MDN reads:_ 
> Workers may, in turn, spawn new workers, **as long as those workers are hosted within the same origin as the parent page**. https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers#web_workers_api

**The most common workaround for this is the following (while others might exist):** 
Instead of spawning the worker with the raw URL, you create a a blob containing the source code which will load the worker through any desired mechanism. When you create an object URL of a blob during runtime, this URL will have the same origin like the website executing the script, even if the executing script has a different origin. Therefore the blob and URL creation can be part of the remote script at the CDN without issues. 

WebPack would need to rewrite the detected expression to this blob syntax like: 

```js
/* WebWorker (proposal with magic comment)  */
new Worker(/* webpackWorkerType: "blob" */ new URL('./worker.js', import.meta.url), { type: 'module' }))
// would become
new Worker(URL.createObjectURL(new Blob([`<generated script>`], { type: 'text/javascript' })), { type: 'module' })

/* Audio Worklet (missing and in discussion how to implement in #11543)  */
await audioContext.audioWorklet.addModule(/* webpackWorkerType: "blob" */ new URL(./worker.js', import.meta.url).href)
// would become 
await audioContext.audioWorklet.addModule(URL.createObjectURL(new Blob([`<generated script>`], { type: 'text/javascript' })));
```

The `<generated script>` will need to be defined and might need certain level of customizations. We might need to control some settings of the blob like the mimetype of the blob which needs to be different for CSS Paint Worklets.

Side note: WebPack currently requires us to pass in a `URL` object to the worker constructor, but the first parameter is [actually expected](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker#parameters) to be a string. But it works currently.  

**The next level of complexity is:** 
* I am the author of a library which is consumed through NPM and not specific to WebPack. 
* The library spawns Web Workers following a `new Worker('@mylib/mylib', import.meta.url)` syntax. In my case I also need to spawn Audio Worklets with the same lib. This is very special because the worker entry point is within the library. In my case it is also the main entry point but it could be also just `@mylib/mylib/worker`.
* People will integrate my library and use it in their applications. They do not have control over the worker creation, the correct code for it needs to come with the library I am developing.  
* Normally mylib would be packed into the vendors.js or a special chunk to ensure the library is only loaded once by the browser. 
* Webpack currently kicks in and generates a worker wrapper script file for the `new Worker(..)` syntax which will then include the vendor.js or the chunk to startup the worker code. 
* Ultimately the output is again hosted like mentioned above. 

**Conclusion:** 
* A compiler setting would require the integrators of my library to change their setup because they are bundling my library into their app. Not a good strategy. There could be conflicts in the settings different libraries might require. Also people need to be guided to configure these settings.
* If WebPack would generate some more code based on settings and syntax to do everything automatic and hidden from the user, I cannot put any customization of the worker for my library into my library. Again the ones using the library would need to take care of this. The more dynamic this customization needs to be, the more unlikely it gets that it will be possible. 
* A separate `__webpack_get_worker_url__` would allow me to unify the whole worker and worklet creation as I can resolve a URL, which allows me to spawn a worker containing my library/worker as entry point. The entry point is the special entry point generated by WebPack for workers. It does not yet support AudioWorklets due to the usage of `importScripts` which is not available there but I'd like to extend this small part.

**Reply by Danielku15** [reply]
An alternative approach would be to add a new magic comment with a built-in code generation for Web Workers like mentioned above. But then we still need separate implementations for all worklets.

**Reply by AprilArcus** [reply]
Excellent summary. Agreed all across.

**Reply by vankop** [reply]
ok. sounds reasonable. not sure yet how this should be implemented..

as workaround I think this setup will work 👇
webpack.config.js
```js
module.exports = [
{ name: "worker", entry: "./src/worker.js", output: { filename: "[name][ext]", path: __dirname } },
{
  name: "app",
  entry: "./src/index",
  dependencies: ["worker"],
  module: {
    rules: { test: /worker\.js$/, type: "asset/source" }
  }
}
];
```
in src/index
```js
async function startWorker() {
    const source = await import("../worker.js"); // builded version
    return new Worker(URL.createObjectURL(new Blob([source], { type: 'text/javascript' })), { type: 'module' })
}
```

**Reply by Danielku15** [reply]
Technically likely yes, but I would have to check it in detail. But I think it is not a user friendly approach. If you are the author of the main app and the worker, you might have these configurations under control and you can tweak it as desired. But library authors will have a tough time guiding people to configure webpack correctly. 

 > not sure yet how this should be implemented..

If we can find a conclusion what would be acceptable for the WebPack maintainers I'd be happy to contribute it. The topic to have proper worker and audio worklet support would be quite a priority topic for me as the users of my library are complaining already quite long about this and it makes it hard to properly integrate my library in modern SPA applications. 

Here some proposals to kick-off the brainstorm: 

1. We go for the __worker_get_worker_url__ which allows devs to decide themselves how to handle it further. Pro: Keeps complexity out of WebPack, Con: An additional module variable which is non standard syntax
2. Introduce a magic comment on the URL syntax at any place: `new URL('./worker.js', import.meta.url)` which would initiate the worker generation logic. Pro: Keeps complexity of handling the URL outside webpack, is not causing errors on other platforms, Con: Might have a performance impact on parsing as we would also inspect all URL creations and need to parse the magic comments there to activate correctly. 
3. Introduce a magic comment which rewrites the worker construction in a fixed practice with `URL.createObjectURL(new Blob(["importScripts(<url>)"], { type: 'text/javascript' }))` Pro: Is not causing errors on other outside webpack usage. Con: Potential memory leak because users cannot revoke the object URL and therefore the blob stays in memory. 

I could think of some more approaches, but they likely would not allow me to solve all my concerns and are not thought through, hence I'm not sharing them.

**Reply by sokra** [reply]
Is it possible to spawn a worker with a DataURI? like that:

``` js
/* WebWorker */
new Worker(new URL('./worker.js', import.meta.url), { type: 'module' }))
// would become
new Worker("data:text/javascript,importScripts(<url>)", { type: 'module' })
```

Maybe we can automatically detect if the worker script is cross-origin based on the base uri.

**Reply by Danielku15** [reply]
Data URIs seem not to be supported like this due the same origin policy of Workers. 

> Data URLs are treated as unique opaque origins by modern browsers, rather than inheriting the origin of the settings object responsible for the navigation. https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs

References: 
https://stackoverflow.com/questions/10343913/how-to-create-a-web-worker-from-a-string
https://lists.whatwg.org/pipermail/whatwg-whatwg.org/2011-March/030878.html
https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs

**Reply by maxammann** [reply]
> Here some proposals to kick-off the brainstorm:

I'm a library author and would like to embed a webworker in the library. I think the only user-friendly way for this is to  include the blob/string within the library. This would essentially mean that there is a bundle within the library.


I would go with option 2, because it is similar to other features in webpack. For example when doing bundle splitting you also include magick comments.

**Reply by Danielku15** [reply]
> I think the only user-friendly way for this is to include the blob/string within the library. 

This works fine for approaches where your worker is an **external** script you want to include. In my case one and the same file of the library acts as API on the browser thread (spawning the worker) **and** it is the script that needs to be launched as worker. 

A blob/string embedded in the library would lead to endless-recrusion as the library would need include itself. 

Blobs or inlined-strings are not solving this use case.

**Reply by veeramarni** [reply]
@Danielku15 I'm interested whether you find any solution since your last comment.

**Reply by Danielku15** [reply]
@veeramarni Yes, for one use case I need, I was able to build a solution that works form me. But it only works when consuming the library I built through NPM. 

In my library code I spawn my worker through:

```js
const worker = new Worker(/* webpackChunkName: "mylib.worker" */ new URL('@mylib/mylib', import.meta.url)
```

The `@mylib/mylib` will be resolved by WebPack to the file holding my library and the related worker code. And by the syntax above the built-in worker support of WebPack kicks in to generate a special bootstrap file which launches webpack and imports my library from a separate chunk (if correctly configured). 

You can see it live here: 

The code which spawns the worker: https://github.com/CoderLine/alphaTab/blob/ca4673373d19759c64f76b98b90cfe2955b86a25/src/Environment.ts#L224
Compiled library: https://cdn.jsdelivr.net/npm/@coderline/alphatab@1.3.0-alpha.369/dist/alphaTab.js
Consumer project: https://github.com/CoderLine/alphaTabSamplesWeb/tree/feature/webpack/src/webpack-js-custom-chunk

I hope it helps you to build something that might work for you too. The goal is to really use the `new Worker(new Url('<something>', import.meta.url))` syntax and you find a way to let `<something>` point to your relevant file/module which WebPack can resolve correctly.

**Reply by veeramarni** [reply]
@Danielku15 I got stuck here. When i import as string

```
const workerString = `import * as alphaTab from '@package/extension/lib/main.worker.index.js'`;
            // our module worksr, in a string
            // const workerBlob = new Blob([workerString]);
            const workerBlob = new Blob([workerString], { type: 'text/javascript' });
            // const workerBlob = new Blob([]);
            const workerUrl = URL.createObjectURL(workerBlob);
            worker = new Worker(workerUrl);
```


It will only import the blob as `import * as alphaTab from '@package/extension/lib/main.worker.index.js` without resolving to the package content.

![image](https://user-images.githubusercontent.com/3382355/190916215-4034ba99-536f-4825-97c4-5442f0d07004.png)

**Reply by Danielku15** [reply]
@veeramarni I think you have misunderstood a bit my message and my code. I have no running version where with some magic strings or combinations you could create an inline-string with a blob. 

Instead of using a blob worker, I use now the [built-in capabilities](https://webpack.js.org/guides/web-workers/) of WebWorkers in WebPack.  The current WebWorker support requires you to launch your worker exactly with a syntax like `new Worker(new URL('./worker.js', import.meta.url));`. 

My pain point in WebPack was that I develop a library which spawns itself as worker and within the library I didn't know what to write for `./worker.js` to make things work after bundling my library with other files (e.g into a vendor.js chunk). 

And this usecase I could solve by 1. detecting whether my library has been bundled with webpack and 2. by assuming that something like '@my/lib/src/worker.js' will resolve within WebPack correctly to my file. 

Combining this together a `new Worker(new URL('@my/lib/src/worker.js', import.meta.url));` will trigger the right stuff in WebPack: 
1. It will create a small bootstrap worker file which integrates correctly with the WebPack ecosystem.
2. It will replace the `new URL(..)` part within my library with this small bootstrap worker file. 
3. It will ensure that the bootstrap worker will import and execute the worker.js file from my library in whatever chunk it is placed launching my library within a worker.

 Here a smaller reference example:

```js
// node_modules/@my/lib/deep-thought.js
export function launchDeepThought(question) {
    const worker = new Worker(new URL('@my/lib/deep-thought.js', import.meta.url));
    worker.postMessage({
      question: question
    });
    worker.onmessage = ({ data: { answer } }) => {
      console.log(answer);
    };
}

if (typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope) {
    self.onmessage = ({ data: { question } }) => {
      self.postMessage({
        answer: 42,
      });
    };
}
        
// app.js
import { launchDeepThought } from '@my/lib/deep-thought.js';
launchDeepThought('The Answer to the Ultimate Question of Life, The Universe, and Everything.');
```

I am the author of a simple library "deep-thought" which can launch itself in a worker. It is consumed through NPM in another app. There  the worker is launched through a `launchDeepThought` function. 

I didn't know what to write to launch my worker because I needed a way to resolve the final URL of the file where my library will be contained (e.g. vendor.js or chunk-deep-thought.js). But with this mechanism of using the library name and the exact syntax that WebPack supports it works. 

I have no solution or approach to let WebPack convert `deep-thought.js` into a Blob or Inline Worker. I let WebPack create a Worker Startup script and let it load my own worker code from a chunk.

Hopefully this makes it a bit clearer to you.

**Reply by veeramarni** [reply]
ok thanks for the explanation. 

I ended up creating webpack config for the worker file and used `asset/resource` to load the file inline. It resolves my problem. In case if anyone is interested in full details. 


webpack config of the package
```
var nodeExternals = require('webpack-node-externals');
var path = require('path');
const webpack = require('webpack');

var webpack_opts = {
  mode: 'development',
  entry: './src/index.ts',
  output: {
    path: path.join(__dirname, 'lib'),
    filename: 'index.js',
    libraryTarget: 'umd',
    globalObject: 'this', //!!!otherwise server fails with ‘window’ is undefined. https://github.com/webpack/webpack/issues/6677
    publicPath: '',
  },
  target: 'web',
  resolve: {
    extensions: ['.ts', '.js', '.mjs', '.jsx', '.tsx', '.graphql', '.graphqls', '.gql'],
    fallback: { "path": require.resolve("path-browserify") },
    alias: {
      // Make it so that 'require' finds the right file.
      fs: "{}"
    }
  },
  devtool: 'source-map',
  module: {
    rules: [{
      test: /\.tsx?$/,
      use: 'ts-loader'
    },
    {
      test: /\.(gql)$/,
      exclude: /node_modules/,
      use: ['graphql-tag/loader']
    },
    {
      test: /\.mjs$/,
      include: /node_modules/,
      type: 'javascript/auto'
    },
    {
      test: /\.graphql?/,
      exclude: /node_modules/,
      use: 'raw-loader',
    },
    {
      resourceQuery: /inline/,
      type: 'asset/source',
    },
    { test: /\.css$/, loader: 'ignore-loader' },
    ]
  },
  // node: false,
  externals: [
    nodeExternals({ modulesDir: "../../../node_modules", allowlist: [/@package\/extension/] }),
    nodeExternals({ allowlist: [/@package\/extension/] })
  ]
};

var webpack_worker = {
  mode: 'development',
  entry: './src/extensions/workers/main.worker.ts',
  output: {
    path: path.join(__dirname, 'lib'),
    filename: 'main.worker.js',
    libraryTarget: 'umd',
    globalObject: 'this', //!!!otherwise server fails with ‘window’ is undefined. https://github.com/webpack/webpack/issues/6677
    publicPath: '',
  },
  target: 'webworker',
  resolve: {
    extensions: ['.ts', '.js', '.mjs', '.jsx', '.tsx', '.graphql', '.graphqls', '.gql'],
    fallback: {
      assert: "assert/",
      buffer: "buffer/",
      fs: false,
      path: require.resolve('path-browserify'),
      http: "stream-http",
      https: "https-browserify",
      os: "os-browserify/browser",
      process: 'process/browser.js',
      stream: 'stream-browserify',
      _stream_duplex: "readable-stream/duplex",
      _stream_passthrough: "readable-stream/passthrough",
      _stream_readable: "readable-stream/readable",
      _stream_transform: "readable-stream/transform",
      _stream_writable: "readable-stream/writable",
      string_decoder: "string_decoder/",
      zlib: 'browserify-zlib',
    },
    alias: {
      // Make it so that 'require' finds the right file.
      fs: "{}"
    }
  },
  plugins: [
    new webpack.ProvidePlugin({
      // this is required for public.config.js
      process: 'process/browser.js',
      Buffer: ['buffer', 'Buffer'],
    }),
    new webpack.DefinePlugin({
      process: {
        APP_ENV: JSON.stringify({
          NODE_ENV: process.env.NODE_ENV
        }),
      }
    }),
  ],
  devtool: 'source-map',
  module: {
    rules: [{
      test: /\.tsx?$/,
      use: 'ts-loader'
    },
    {
      test: /\.(gql)$/,
      exclude: /node_modules/,
      use: ['graphql-tag/loader']
    },
    {
      test: /\.mjs$/,
      include: /node_modules/,
      type: 'javascript/auto'
    },
    {
      test: /\.graphql?/,
      exclude: /node_modules/,
      use: 'raw-loader',
    },

    { test: /\.css$/, loader: 'ignore-loader' },
    ]
  },
  // node: false,
};

module.exports = [webpack_worker, webpack_opts];
```

And for worker, i do something like this

```

            const workerString = require('@package/extension/lib/main.worker.js?inline');
            const workerBlob: Blob = new Blob([workerString], { type: 'text/javascript' });
            const workerUrl = URL.createObjectURL(workerBlob);
            worker = new Worker(workerUrl);
```

**Comment by openHacking** [comment]
I need this.I think it is very import feature.

**Reply by maxammann** [reply]
If you are developing a library I suggest to use esbuild with https://github.com/mitschabaude/esbuild-plugin-inline-worker

It's simple and works flawlessly.

**Comment by whitphx** [comment]
I could bypassed the cross-origin problem with [this trick](https://github.com/webpack/webpack/discussions/14648#discussioncomment-1589272) instead of inlining the worker code.
While it was introduced as a workaround for the CORS problem, it worked for this problem too.

This is just a workaround, but I hope it helps someone for now. And huge thanks to the original author, @piotr-oles.

I would just copy and paste the code for reference:
```typescript
export class CorsWorker {
    private readonly worker: Worker;

    constructor(url: string | URL) {
        const objectURL = URL.createObjectURL(
            new Blob([`importScripts(${JSON.stringify(url.toString())});`], {
                type: 'application/javascript'
            })
        );
        this.worker = new Worker(objectURL);
        URL.revokeObjectURL(objectURL);
    }

    getWorker() {
        return this.worker;
    }
}
```

```typescript
import { CorsWorker as Worker } from './cors-worker';

// aliasing CorsWorker to Worker makes it statically analyzable
const corsWorker = new Worker(new URL('../worker', import.meta.url));
// tada!
const worker = corsWorker.getWorker();
```

The idea is using a wrapper class `CorsWorker` that initializes a worker from blob and load the actual worker implementation via `importScripts`, and importing the wrapper class as `Worker` alias in order to use the special syntax `new Worker(new URL('./worker.js', import.meta.url))` and inject the wrapper class into it.

---
FYI: This is the code in my actual project.
The definition of the wrapper class.
https://github.com/whitphx/stlite/blob/35bb1be421e400313a952400d9e76b0c72358256/packages/kernel/src/cross-origin-worker.ts

The caller file where the wrapper class is imported as `Worker` alias and the special syntax for Webpack's Worker feature is used.
https://github.com/whitphx/stlite/blob/35bb1be421e400313a952400d9e76b0c72358256/packages/kernel/src/kernel.ts#L6
https://github.com/whitphx/stlite/blob/35bb1be421e400313a952400d9e76b0c72358256/packages/kernel/src/kernel.ts#L102-L104

**Reply by augustVino** [reply]
```new Blob([`importScripts(${JSON.stringify(url.toString())});`], { type: 'application/javascript' })```
This way really solved my problem, but I don't understand where the importScripts function in the code is defined? How does it work? Can anyone help me solve my doubts?

**Reply by whitphx** [reply]
It's a built-in function: https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts

**Comment by ShuZhong** [comment]
In latest webpack, that's a `workerPublicPath` to override `publicPath` in worker use.
we can set `publicPath: 'https://www.cdn.com` and `workerPublicPath: '/'` to solve CORS errors.

https://webpack.js.org/configuration/output/#outputworkerpublicpath
